keps/sig-apps/20190412-pdb-for-scale-subresource.md

```diff
@@ -0,0 +1,86 @@
+---
+title: support-scale-subresource-in-pdbs
+authors:
+  - "@mortent"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-scheduling
+  - sig-autoscaling
+reviewers:
+  - "@kow3ns"
+  - "@janetkuo"
+approvers:
+  - "@kow3ns"
+  - "@janetkuo"
+editor: TBD
+creation-date: 2019-04-12
+last-updated: 2019-04-12
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Support scale subresource in PDBs
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+
+## Summary
+
+[Pod Disruption Budget (PDB)](https://kubernetes.io/docs/tasks/run-application/configure-pdb/)
+is a Kubernetes API that limits the number of pods of a collection that are down simultaneously from voluntary disruptions. PDBs allows a user to specify the allowed disruption through either min available or max unavailable number of pods. In order to support PDBs where max number of unavailable pods is set, the PDB controller needs to be able to look up the desired number of replicas. It does this by looking at the controller(s) (as specified by the owner ref) for the pods covered by the PDB. Currently only four of the basic workload controllers are supported for PDBs, namely Deployment, StatefulSet, ReplicaSet and ReplicationController. 
+
+The scale subresource allows any resource to specify its desired number of replicas and a generic way to look up this information. This document lays out a plan to use the scale subresource to allow setting PDBs for any resource implementing the scale subresource.
+
```

Comment:
PDB is specific for Pods. However, not all custom resources that support /scale subresource use # of replicas for # of Pods. It could be something other than Pods.

Reply:
So if this happens, then the pods will not reference the custom resource and whichever controller is referenced by the pods covered by the PDB will be used. 

The scenario that isn't supported for custom resources, is what currently happens for deployments. The PDB controller reads the scale from the Deployment referenced by the controllerRef on a ReplicaSet, rather than directly from the ReplicaSet. I haven't found a way to support this with custom resources and I think it is a limitation we can live with.